### PR TITLE
Add --rebase flag for arc diff to enable rebase before creating the diff

### PR DESCRIPTION
**Pull request by xiciluan**
### Summary
Original proposal [here](https://docs.google.com/document/d/1EBIkpzG0r6-UK6uvZhwZeB-3SheqZEMDT8UPNASvHnA/edit#heading=h.s3pp339uvtkp).

This PR will add a `--rebase` flag for `arc diff`.  
To enable rebase before creating diff, run `arc diff --rebase true`; otherwise, run `arc diff --rebase false` or just `arc diff`. Default value for is `false` for now. Will switch to `true` in the future.

### Testing
After making some dummy update on a local branch, if there's no merge conflicts between local branch and stable, console output:
```
$ arc diff --rebase true
Running arc rebase...
 SHELL ALIAS  arc rebase -> $ eval '$(git rev-parse --show-toplevel)/arcanist/aliases/rebase.sh'
remote: Enumerating objects: 564, done.
remote: Counting objects: 100% (564/564), done.
remote: Compressing objects: 100% (120/120), done.
remote: Total 353 (delta 276), reused 306 (delta 231), pack-reused 0
Receiving objects: 100% (353/353), 83.77 KiB | 4.93 MiB/s, done.
Resolving deltas: 100% (276/276), completed with 193 local objects.
From github.com:robinhoodmarkets/rh
 * branch                    master     -> FETCH_HEAD
 * branch                    stable     -> FETCH_HEAD
   45c17b5facb..87c3de84999  master     -> origin/master
   fd80e976a8d..6e809ff7eba  stable     -> origin/stable
Successfully rebased and updated refs/heads/tmp2.
Returned with status 0 and outputs:
``` 
If there are merge conflicts between the two branch, there will be prompt to determine if to continue creating the diff. 
```
$ arc diff --rebase true
Running arc rebase...
 SHELL ALIAS  arc rebase -> $ eval '$(git rev-parse --show-toplevel)/arcanist/aliases/rebase.sh'
remote: Enumerating objects: 25, done.
remote: Counting objects: 100% (25/25), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 14 (delta 12), reused 13 (delta 11), pack-reused 0
Unpacking objects: 100% (14/14), 1.68 KiB | 95.00 KiB/s, done.
From github.com:robinhoodmarkets/rh
 * branch                    master     -> FETCH_HEAD
 * branch                    stable     -> FETCH_HEAD
   732a595c804..186e1068722  master     -> origin/master
   25bcf1c100f..aa6a70f9184  stable     -> origin/stable
error: could not apply 142a0e0814b... create a conflict
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 142a0e0814b... create a conflict
Returned with status 1 and outputs:
Auto-merging sickle/integration_tests/roles/sickle/supervisord.conf
CONFLICT (content): Merge conflict in sickle/integration_tests/roles/sickle/supervisord.conf


    arc rebase failed, do you want to continue creating the diff? [y/N] 
```
If yes, 
```
arc rebase failed, do you want to continue creating the diff? [y/N] y

Running git rebase --abort
```
If no, 
```
    arc rebase failed, do you want to continue creating the diff? [y/N] h


    arc rebase failed, do you want to continue creating the diff? [y/N] n

Stop creating diff...
```

If there's no wifi or VPN connected, same exception will be thrown for `arc diff` and `arc diff --rebase true`.

### CODE REVIEW

File: src/workflow/ArcanistDiffWorkflow.php [hunk 1]
```diff
@@ -344,12 +348,44 @@ public function isRawDiffSource() {
     return $this->getArgument('raw') || $this->getArgument('raw-command');
   }
 
+  private function rebaseToStable() {
```
Comment (naturalwarren) [inline src/workflow/ArcanistDiffWorkflow.php:351]: I think this whole function can be replaced inline at the call site. The call site would then read:

```
$is_rebasing = trim(strtolower($this->getArgument('rebase'))) == 'true'
if ($is_rebasing) {
  echo "Running arc rebase... \n";
  ...
}
```

**Review by naturalwarren — APPROVED** [review]
One comment, please address prior to releasing.

Love how this is off by default to begin with. Let's give it a whirl.

**Review by naturalwarren — APPROVED** [review]

**Comment by alexeagle** [comment]
nit: 
> arc rebase failed, do you want to continue creating the diff
I think users may think "wait, if it failed I'd better fix it before I create the diff or else someone will yell at me"

maybe
arc rebase failed. You can skip the rebase and continue creating the diff with the original base commit, then deal with the merge conflicts later. Skip rebase? [y/N]

File: src/workflow/ArcanistDiffWorkflow.php [hunk 2]
```diff
@@ -344,12 +348,52 @@ public function isRawDiffSource() {
...
+      exec('arc rebase', $outputs, $retval);
+      echo "Returned with status $retval and outputs:\n";
+      foreach ($outputs as $output) {
+        echo "$output\n";
+      }
+      if ($retval == 1) {
+        $prompt = "arc rebase failed. You can skip the rebase and continue creating the diff with the original base commit, then deal with the merge conflicts later. Skip rebase?";
+        $continue_diff = phutil_console_confirm($prompt, $default_no = true);
+        if (!$continue_diff) {
+          echo "Stop creating diff... \n";
```
Comment (txchen) [inline src/workflow/ArcanistDiffWorkflow.php:376]: If user choose 'no', I suppose user need to manually continue the rebase and resolve conflicts. Shall we make it clear here to users what they should do next?
Also, if rebase is still in-progress, and we run arc diff -> arc rebase again, what will happen?

Comment (xiciluan) [reply src/workflow/ArcanistDiffWorkflow.php:376]: When running `arc rebase` and failed, console output will tell to resolve the conflict and run `git rebase --continue/abort/skip`. Sure we can clarify it here saying "please follow the above instructions to resolve the conflicts".

If the rebase is still in-progress and we run `arc rebase` again, console output will be "rebase already in progress, and give instructions for using git rebase --rebase/abort/skip to resolve the issue".

Comment (txchen) [reply src/workflow/ArcanistDiffWorkflow.php:376]: Thanks for then information! I think it is a good UX. Shall I merge the PR now?

**Review by txchen — DISMISSED** [review]
Approved.

**Review by txchen — APPROVED** [review]

**Pull request merged**